### DISCUSSION TITLE
Component appears twice

### DISCUSSION THREAD

**Discussion by Zonalds** [vercel/next.js#36315] [Help] [answered]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_X86_64
    Binaries:
      Node: 16.13.1
      npm: 8.5.4
      Yarn: 1.22.17
      pnpm: N/A
    Relevant packages:
      next: 12.1.5
      react: 18.0.0
      react-dom: 18.0.0

warn  - Latest canary version not detected, detected: "12.1.5", newest: "12.1.6-canary.4".
        Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
        Read more - https://nextjs.org/docs/messages/opening-an-issue

### What browser are you using? (if relevant)

Chrome & Brave

### How are you deploying your application? (if relevant)

Vercel

### What happened?

NextJs renders components twice. The component appears twice on the browser though only one is updated on useState. I have tried nearly everything including using dynamic import.

![image](https://user-images.githubusercontent.com/50281058/164235293-971eb58e-52e1-4635-98ae-cb671a15150e.png)

This only happens NextJS, CRA works fine.  

### Version & 3P packages

@carbon/charts-react@^0.57.0

### Data & options used

```js
import React from "react";
// import ReactDOM from "react-dom";
import { SimpleBarChart } from "@carbon/charts-react";
import "@carbon/charts/styles.css";
// Or
// import "@carbon/charts/styles/styles.scss";

// IBM Plex should either be imported in your project by using Carbon
// or consumed manually through an import
// import "./plex-and-carbon-components.css";

class App extends React.Component {
    state = {
        data: [
            {
                "group": "Qty",
                "value": 65000
            },
            {
                "group": "More",
                "value": 29123
            },
            {
                "group": "Sold",
                "value": 35213
            },
            {
                "group": "Restocking",
                "value": 51213
            },
            {
                "group": "Misc",
                "value": 16932
            }
        ],
        options: {
            "title": "Horizontal simple bar (discrete)",
            "axes": {
                "left": {
                    "mapsTo": "group",
                    "scaleType": "labels"
                },
                "bottom": {
                    "mapsTo": "value"
                }
            },
            "height": "400px"
        }
    };

    render = () => (
        <SimpleBarChart
            data={this.state.data}
            options={this.state.options}>
        </SimpleBarChart>
    );
}

export default App;
```

### Interesting discoveries
```Development
This only happens in development. On production instances, the charts are rendered correctly - appear once.

SSR doesn't seem to have any effect on it. The chart appears twice regardless of whether the SSR is turned false or on. No regular dynamic import configuration works so far.
```

### Relevant log output

```js
No log output.
```

### Expected Behavior

The component should appear only once

### To Reproduce

Just post this code in any page route and run

### Data & options used

```js
import React from "react";
// import ReactDOM from "react-dom";
import { SimpleBarChart } from "@carbon/charts-react";
import "@carbon/charts/styles.css";
// Or
// import "@carbon/charts/styles/styles.scss";

// IBM Plex should either be imported in your project by using Carbon
// or consumed manually through an import
// import "./plex-and-carbon-components.css";

class App extends React.Component {
    state = {
        data: [
            {
                "group": "Qty",
                "value": 65000
            },
            {
                "group": "More",
                "value": 29123
            },
            {
                "group": "Sold",
                "value": 35213
            },
            {
                "group": "Restocking",
                "value": 51213
            },
            {
                "group": "Misc",
                "value": 16932
            }
        ],
        options: {
            "title": "Horizontal simple bar (discrete)",
            "axes": {
                "left": {
                    "mapsTo": "group",
                    "scaleType": "labels"
                },
                "bottom": {
                    "mapsTo": "value"
                }
            },
            "height": "400px"
        }
    };

    render = () => (
        <SimpleBarChart
            data={this.state.data}
            options={this.state.options}>
        </SimpleBarChart>
    );
}

export default App;
```

Be sure to install @carbon/charts-react.

**Comment by timneutkens** [comment]
It can be caused one of these two:
- Your app is mismatching rendered html during hydration, in that case the console would be full of messages saying hydration mismatched and where it mismatched
- You have strict mode enabled and are using React 18 which changed strict mode semantics to catch concurrent rendering issues. You can read more about that here: https://reactjs.org/blog/2022/03/29/react-v18.html#gradually-adopting-concurrent-features. That would indicate the component you're using is not compatible with strict mode and you'll want to file an issue on the respective repository.

**Reply by Zonalds** [reply]
Thank you for your kind response. I don't fully understand this so I will move to use a different library. 


So far I have tested multiple charting libraries like bizchart, and chartkick but have this issue where they blink and disappear. They appear during load and quickly disappear.

Is there an example implementation of any of this with NextJs or a guide on solving this?

**Comment by icyJoseph** [comment, marked as answer]
To follow up on the first answer.

There's something called Strict Mode, which one can use during development mode. The aim is to pin point areas where pure functions are causing unwanted behaviour.

In your `next.config.js` file, change `reactStrictMode` to false, and this error will disappear.

However, the charting library must be doing something "not-pure", which causes this behaviour. 

In Create React App or Vite, you can trigger the same by importing StrictMode from `react`, and wrapping your application with it:

```tsx
import { StrictMode } from "react";
import { createRoot } from "react-dom/client";

import App from "./App";

const rootElement = document.getElementById("root");
const root = createRoot(rootElement);

root.render(
  <StrictMode>
    <App />
  </StrictMode>
);

```

**Reply by Zonalds** [reply]
Thank you very much. That solved all issues including the disappearing component.

**Reply by icyJoseph** [reply]
Hi,

For completeness sake, I went to the carbon Github page, and found the simple bar component.

We can now make absolute sense of the issue.

When they mount the chart, they assign an `SBC`(internal stuff) object to `this.charts`, in `componentDidMount`, and output a `div` to the `DOM`.

https://github.com/carbon-design-system/carbon-charts/blob/master/packages/react/src/bar-chart-simple.tsx#L13

The React 18 Strict Mode, **in development mode**, simulates an immediate `unmount` and  re-`mount`, immediately.

>  With Strict Mode starting in React 18, whenever a component mounts in development, React will simulate immediately unmounting and remounting the component

But they never clean up the `SBC` object, which I believe keeps the chart up in the DOM. They sort of expected that to happen when the entire component un-mounts, but the Strict Mode behaviour forces the library to again output a `div` to the `DOM`. This is why we see duplicates.

The library has to implement `componentWillUnMount` where they should take care of cleaning up the `DOM`.

**Reply by SyedHanifShah** [reply]
thanks a lot, changing reactStrictMode to false solved the issue.
